drivers/src/pci/config_space.rs

```diff
@@ -0,0 +1,296 @@
+        }
+
+        // Checks for a `bus-range` property specifying a start bus.
+        let start_bus_index = pci_node
+            .props()
+            .find(|p| p.name() == "bus-range")
+            .and_then(|props| props.value_u32().next())
+            .unwrap_or(0);
+        let start_bus = start_bus_index
+            .try_into()
+            .map_err(|_| Error::StartBusInvalid(start_bus_index))?;
+
+        unsafe {
+            // Safety: Have to trust that the device tree points to valid PCI space.
+            // Any overlaps will be caught by `add_mmio_region` and the error will be propagated.
+            mem_map
+                .add_mmio_region(
+                    DeviceMemType::PciRoot,
+                    RawAddr::from(config_base),
+                    config_size,
+                )
+                .map_err(Error::InvalidMmioRegion)?;
+        }
+
+        Ok(Self {
+            config_base,
+            config_size,
+            // TODO create a config space per segment
+            segment: Segment::default(),
+            start_bus,
+        })
+    }
+
+    /// Returns an iterator across the top-level buses.
+    pub fn busses(&self) -> HostControllersIter {
+        let header_addr = Address::new(
+            self.segment,
+            self.start_bus,
+            Device::default(),
+            Function::default(),
+        );
+
+        HostControllersIter {
+            config_space: self,
+            segment: self.segment,
+            fn_range: self.function_scan_range(header_addr),
+        }
+    }
+
+    /// Gets a Header object at the given address if it exists.
+    pub fn header_for(&self, address: Address) -> Option<Header> {
+        let offset = self.header_offset(address)?;
+        // Safety: config_base is guaranteed to be uniquely owned PCI memory by construction and the
+        // range of memory used below is within the owned range as checked.
+        unsafe {
+            if offset.checked_add(PCIE_FUNCTION_HEADER_LEN)? > self.config_size {
+                return None;
+            }
+            Header::new(address, (self.config_base.bits() + offset) as *mut Le32)
+        }
+    }
+
+    // Returns the range of headers to check based on if this is a multi-function device.
+    // If no header present: 0..0 don't scan anything.
+    // If the header is present and multi-function: 0..8 - Check all the possible headers.
+    // If the header is present and not multi-function: 0..1
+    fn function_scan_range(&self, header_addr: Address) -> Range<u32> {
+        // Only applicable to first function of a device.
+        if header_addr.function() != Function::default() {
+            return 0..0;
+        }
+        self.header_for(header_addr)
+            .map(|header| if header.multi_function() { 0..8 } else { 0..1 })
+            .unwrap_or(0..0)
+    }
+
+    // Returns the offset of the given address within a PciConfigSpace starting at
+    // `config_bus_start`.
```

Comment:
"... of the given address within this PciConfigSpace" ?

Reply:
yes, applied. Thanks!